File: bg-housing-rollups.r
## BLOCK GROUP TO GEOGRAPHIC ROLLUPS
## CORE ADAPTED FROM 'zcta-to-county.r'

library(duckdb)
library(data.table)
library(DBI)
library(dplyr)
#install.packages("reactable")
library(reactable)
library(plotly)
library(classInt)
library(RColorBrewer)
library(dotenv)
library(sf)

setwd("/home/joel")

dotenv::load_dot_env()

### DUCKDB CONNECTION

conh <- dbConnect(duckdb('./data/housing.duckdb'))
dbListTables(conh)

congref <- dbConnect(duckdb('./data/georeference.duckdb'))
dbListTables(congref)

congeo <- dbConnect(
  duckdb::duckdb(), 
  dbdir = "./data/spatial_storage.duckdb",
  read_only = FALSE,
  extensions = c("spatial")
)

dbListTables(congeo)

metadata <- dbGetQuery(conmeta, "select * from column_metadata")
tables <- dbGetQuery(conmeta, "select * from table_summary")

# load block groups

hu_block_group <- dbGetQuery(conh, "select * from hu_block_group")

# match block groups to cbsa via county

county_to_cbsa <- fread("./data/county-cbsa-lookup.csv", 
                        colClasses = "character",
                        encoding = "Latin-1")

## write to georeference duckdb

dbWriteTable(congref, "county_to_cbsa", county_to_cbsa, overwrite = TRUE)

## get comparison geographies for index computations
## COUNTY, CBSA, STATE, US 

hu_block_group <- as.data.table(dbGetQuery(conh, "select * from hu_block_group"))
hu_county <- as.data.table(dbGetQuery(conh, "select * from hu_county"))
hu_cbsa <- as.data.table(dbGetQuery(conh, "select * from hu_cbsa"))
hu_state <- as.data.table(dbGetQuery(conh, "select * from hu_state"))
hu_us <- as.data.table(dbGetQuery(conh, "select * from hu_us"))

#################################### CBSA LEVEL BG ROLLUP #############

county_to_cbsa <- as.data.table(dbGetQuery(congref, "select * from county_to_cbsa"))

## compute county fips
hu_block_group[, county_fips := substr(block_group, 1, 5)]

hu_block_group_cbsa <- merge(hu_block_group, county_to_cbsa, by.x = "county_fips", by.y = "county", all.x = TRUE)

hu_block_group_cbsa <- as.data.table(
  merge(hu_block_group_cbsa, hu_cbsa, by.x = "cbsa23", by.y = "cbsa23", all.x = TRUE)
)

############### RELATIVE INDEX COMPUTATIONS  -- CBSA LEVEL

# hgi_* indexes (housing growth index)
hu_block_group_cbsa[, idx_cbsa_hgi_20_apr_24_jul := (hgi_20_apr_24_jul.x - 1) / (hgi_20_apr_24_jul.y - 1) * 100]
hu_block_group_cbsa[, idx_cbsa_hgi_24_jul_25_jul := (hgi_24_jul_25_jul.x - 1) / (hgi_24_jul_25_jul.y - 1) * 100]
hu_block_group_cbsa[, idx_cbsa_hgi_24_jul_25_nov := (hgi_24_jul_25_nov.x - 1) / (hgi_24_jul_25_nov.y - 1) * 100]
hu_block_group_cbsa[, idx_cbsa_hgi_25_jul_25_nov := (hgi_25_jul_25_nov.x - 1) / (hgi_25_jul_25_nov.y - 1) * 100]
hu_block_group_cbsa[, idx_cbsa_hgi_20_apr_25_jul := (hgi_20_apr_25_jul.x - 1) / (hgi_20_apr_25_jul.y - 1) * 100]
hu_block_group_cbsa[, idx_cbsa_hgi_20_apr_25_nov := (hgi_20_apr_25_nov.x - 1) / (hgi_20_apr_25_nov.y - 1) * 100]

# cagr_* indexes (compound annual growth rate)
hu_block_group_cbsa[, idx_cbsa_cagr_20_apr_24_jul := (cagr_20_apr_24_jul.x) / (cagr_20_apr_24_jul.y) * 100]
hu_block_group_cbsa[, idx_cbsa_cagr_24_jul_25_nov := (cagr_24_jul_25_nov.x) / (cagr_24_jul_25_nov.y) * 100]
hu_block_group_cbsa[, idx_cbsa_cagr_25_jul_25_nov := (cagr_25_jul_25_nov.x) / (cagr_25_jul_25_nov.y) * 100]
hu_block_group_cbsa[, idx_cbsa_cagr_20_apr_25_jul := (cagr_20_apr_25_jul.x) / (cagr_20_apr_25_jul.y) * 100]
hu_block_group_cbsa[, idx_cbsa_cagr_20_apr_25_nov := (cagr_20_apr_25_nov.x) / (cagr_20_apr_25_nov.y) * 100]

# agr_* indexes (annual growth rate)
hu_block_group_cbsa[, idx_cbsa_agr_24_jul_25_jul := (agr_24_jul_25_jul.x) / (agr_24_jul_25_jul.y) * 100]

## NULL CBSA ARE CONNECTICUT BLOCK GROUPS; CBSA DEFINITIONS BASED ON NEW PLANNING REGION
## COMPONENTS ARE NOT YET DEFINED; WE ARE USING NEW PLANNING REGION IDS IN OUR BLOCK GROUP
## IDENTIFIERS; WE WILL NEED A PLANNING REGION TO CBSA CROSSWALK TO HANDLE THESE; THESE ROWS 
## WOULD REPLACE COUNTY-BASED DEFINITIONS

hu_block_group_cbsa <- hu_block_group_cbsa[!is.na(block_group), ]

hu_block_group_cbsa[is.na(cbsa23), cbsa23 := "CT000"]

## TRIM FILE

hu_block_group_cbsa <- hu_block_group_cbsa[, .(
  block_group,
  cbsa23,
  idx_cbsa_hgi_20_apr_24_jul,
  idx_cbsa_hgi_24_jul_25_jul,
  idx_cbsa_hgi_24_jul_25_nov,
  idx_cbsa_hgi_25_jul_25_nov,
  idx_cbsa_hgi_20_apr_25_jul,
  idx_cbsa_hgi_20_apr_25_nov,
  idx_cbsa_cagr_20_apr_24_jul,
  idx_cbsa_cagr_24_jul_25_nov,
  idx_cbsa_cagr_25_jul_25_nov,
  idx_cbsa_cagr_20_apr_25_jul,
  idx_cbsa_cagr_20_apr_25_nov,
  idx_cbsa_agr_24_jul_25_jul
)]

#################################### COUNTY LEVEL BG ROLLUP #############

hu_block_group_county <- as.data.table(
  merge(hu_block_group, hu_county, by.x = "county_fips", by.y = "co_fips", all.x = TRUE)
)

################# RELATIVE INDEX COMPUTATIONS  -- COUNTY LEVEL

# hgi_* indexes (housing growth index)
hu_block_group_county[, idx_county_hgi_20_apr_24_jul := (hgi_20_apr_24_jul.x - 1) / (hgi_20_apr_24_jul.y - 1) * 100]
hu_block_group_county[, idx_county_hgi_24_jul_25_jul := (hgi_24_jul_25_jul.x - 1) / (hgi_24_jul_25_jul.y - 1) * 100]
hu_block_group_county[, idx_county_hgi_24_jul_25_nov := (hgi_24_jul_25_nov.x - 1) / (hgi_24_jul_25_nov.y - 1) * 100]
hu_block_group_county[, idx_county_hgi_25_jul_25_nov := (hgi_25_jul_25_nov.x - 1) / (hgi_25_jul_25_nov.y - 1) * 100]
hu_block_group_county[, idx_county_hgi_20_apr_25_jul := (hgi_20_apr_25_jul.x - 1) / (hgi_20_apr_25_jul.y - 1) * 100]
hu_block_group_county[, idx_county_hgi_20_apr_25_nov := (hgi_20_apr_25_nov.x - 1) / (hgi_20_apr_25_nov.y - 1) * 100]

# cagr_* indexes (compound annual growth rate)
hu_block_group_county[, idx_county_cagr_20_apr_24_jul := (cagr_20_apr_24_jul.x) / (cagr_20_apr_24_jul.y) * 100]
hu_block_group_county[, idx_county_cagr_24_jul_25_nov := (cagr_24_jul_25_nov.x) / (cagr_24_jul_25_nov.y) * 100]
hu_block_group_county[, idx_county_cagr_25_jul_25_nov := (cagr_25_jul_25_nov.x) / (cagr_25_jul_25_nov.y) * 100]
hu_block_group_county[, idx_county_cagr_20_apr_25_jul := (cagr_20_apr_25_jul.x) / (cagr_20_apr_25_jul.y) * 100]
hu_block_group_county[, idx_county_cagr_20_apr_25_nov := (cagr_20_apr_25_nov.x) / (cagr_20_apr_25_nov.y) * 100]

# agr_* indexes (annual growth rate)
hu_block_group_county[, idx_county_agr_24_jul_25_jul := (agr_24_jul_25_jul.x) / (agr_24_jul_25_jul.y) * 100]

hu_block_group_county <- hu_block_group_county[!is.na(block_group), ]

## TRIM FILE

hu_block_group_county <- hu_block_group_county[, .(
  block_group,
  county_fips,
  idx_county_hgi_20_apr_24_jul,
  idx_county_hgi_24_jul_25_jul,
  idx_county_hgi_24_jul_25_nov,
  idx_county_hgi_25_jul_25_nov,
  idx_county_hgi_20_apr_25_jul,
  idx_county_hgi_20_apr_25_nov,
  idx_county_cagr_20_apr_24_jul,
  idx_county_cagr_24_jul_25_nov,
  idx_county_cagr_25_jul_25_nov,
  idx_county_cagr_20_apr_25_jul,
  idx_county_cagr_20_apr_25_nov,
  idx_county_agr_24_jul_25_jul
)]

################################## STATE LEVEL BG ROLLUP #############

hu_block_group[, state_fips := substr(block_group, 1, 2)]

hu_block_group_state <- as.data.table(
  merge(hu_block_group, hu_state, by.x = "state_fips", by.y = "state_code", all.x = TRUE)
)

############### RELATIVE INDEX COMPUTATIONS  -- STATE LEVEL

# hgi_* indexes (housing growth index)
hu_block_group_state[, idx_state_hgi_20_apr_24_jul := (hgi_20_apr_24_jul.x - 1) / (hgi_20_apr_24_jul.y - 1) * 100]
hu_block_group_state[, idx_state_hgi_24_jul_25_jul := (hgi_24_jul_25_jul.x - 1) / (hgi_24_jul_25_jul.y - 1) * 100]
hu_block_group_state[, idx_state_hgi_24_jul_25_nov := (hgi_24_jul_25_nov.x - 1) / (hgi_24_jul_25_nov.y - 1) * 100]
hu_block_group_state[, idx_state_hgi_25_jul_25_nov := (hgi_25_jul_25_nov.x - 1) / (hgi_25_jul_25_nov.y - 1) * 100]
hu_block_group_state[, idx_state_hgi_20_apr_25_jul := (hgi_20_apr_25_jul.x - 1) / (hgi_20_apr_25_jul.y - 1) * 100]
hu_block_group_state[, idx_state_hgi_20_apr_25_nov := (hgi_20_apr_25_nov.x - 1) / (hgi_20_apr_25_nov.y - 1) * 100]

# cagr_* indexes (compound annual growth rate)
hu_block_group_state[, idx_state_cagr_20_apr_24_jul := (cagr_20_apr_24_jul.x) / (cagr_20_apr_24_jul.y) * 100]
hu_block_group_state[, idx_state_cagr_24_jul_25_nov := (cagr_24_jul_25_nov.x) / (cagr_24_jul_25_nov.y) * 100]
hu_block_group_state[, idx_state_cagr_25_jul_25_nov := (cagr_25_jul_25_nov.x) / (cagr_25_jul_25_nov.y) * 100]
hu_block_group_state[, idx_state_cagr_20_apr_25_jul := (cagr_20_apr_25_jul.x) / (cagr_20_apr_25_jul.y) * 100]
hu_block_group_state[, idx_state_cagr_20_apr_25_nov := (cagr_20_apr_25_nov.x) / (cagr_20_apr_25_nov.y) * 100]

# agr_* indexes (annual growth rate)
hu_block_group_state[, idx_state_agr_24_jul_25_jul := (agr_24_jul_25_jul.x) / (agr_24_jul_25_jul.y) * 100]

hu_block_group_state <- hu_block_group_state[!is.na(block_group), ]

## TRIM FILE

hu_block_group_state <- hu_block_group_state[, .(
  block_group,
  state_fips,
  idx_state_hgi_20_apr_24_jul,
  idx_state_hgi_24_jul_25_jul,
  idx_state_hgi_24_jul_25_nov,
  idx_state_hgi_25_jul_25_nov,
  idx_state_hgi_20_apr_25_jul,
  idx_state_hgi_20_apr_25_nov,
  idx_state_cagr_20_apr_24_jul,
  idx_state_cagr_24_jul_25_nov,
  idx_state_cagr_25_jul_25_nov,
  idx_state_cagr_20_apr_25_jul,
  idx_state_cagr_20_apr_25_nov,
  idx_state_agr_24_jul_25_jul
)]

################################## US LEVEL BG ROLLUP #############

hu_block_group[, matchid := 1]
hu_us[, matchid := 1]

hu_block_group_us <- as.data.table(
  merge(hu_block_group, hu_us, by.x = "matchid",  by.y = "matchid", all = TRUE)
)

############# RELATIVE INDEX COMPUTATIONS  -- US LEVEL

# hgi_* indexes (housing growth index)
hu_block_group_us[, idx_us_hgi_20_apr_24_jul := (hgi_20_apr_24_jul.x - 1) / (hgi_20_apr_24_jul.y - 1) * 100]
hu_block_group_us[, idx_us_hgi_24_jul_25_jul := (hgi_24_jul_25_jul.x - 1) / (hgi_24_jul_25_jul.y - 1) * 100]
hu_block_group_us[, idx_us_hgi_24_jul_25_nov := (hgi_24_jul_25_nov.x - 1) / (hgi_24_jul_25_nov.y - 1) * 100]
hu_block_group_us[, idx_us_hgi_25_jul_25_nov := (hgi_25_jul_25_nov.x - 1) / (hgi_25_jul_25_nov.y - 1) * 100]
hu_block_group_us[, idx_us_hgi_20_apr_25_jul := (hgi_20_apr_25_jul.x - 1) / (hgi_20_apr_25_jul.y - 1) * 100]
hu_block_group_us[, idx_us_hgi_20_apr_25_nov := (hgi_20_apr_25_nov.x - 1) / (hgi_20_apr_25_nov.y - 1) * 100]

# cagr_* indexes (compound annual growth rate)
hu_block_group_us[, idx_us_cagr_20_apr_24_jul := (cagr_20_apr_24_jul.x) / (cagr_20_apr_24_jul.y) * 100]
hu_block_group_us[, idx_us_cagr_24_jul_25_nov := (cagr_24_jul_25_nov.x) / (cagr_24_jul_25_nov.y) * 100]
hu_block_group_us[, idx_us_cagr_25_jul_25_nov := (cagr_25_jul_25_nov.x) / (cagr_25_jul_25_nov.y) * 100]
hu_block_group_us[, idx_us_cagr_20_apr_25_jul := (cagr_20_apr_25_jul.x) / (cagr_20_apr_25_jul.y) * 100]
hu_block_group_us[, idx_us_cagr_20_apr_25_nov := (cagr_20_apr_25_nov.x) / (cagr_20_apr_25_nov.y) * 100]

# agr_* indexes (annual growth rate)
hu_block_group_us[, idx_us_agr_24_jul_25_jul := (agr_24_jul_25_jul.x) / (agr_24_jul_25_jul.y) * 100]


### COMPUTE NATIONAL PERCENTILES FOR BLOCK GROUPS

# Percentiles for hgi_* metrics
hu_block_group_us[, pctl_us_hgi_20_apr_24_jul := as.integer(
  ceiling(frank(hgi_20_apr_24_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_hgi_24_jul_25_jul := as.integer(
  ceiling(frank(hgi_24_jul_25_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_hgi_24_jul_25_nov := as.integer(
  ceiling(frank(hgi_24_jul_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_hgi_25_jul_25_nov := as.integer(
  ceiling(frank(hgi_25_jul_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_hgi_20_apr_25_jul := as.integer(
  ceiling(frank(hgi_20_apr_25_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_hgi_20_apr_25_nov := as.integer(
  ceiling(frank(hgi_20_apr_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]

# Percentiles for cagr_* metrics
hu_block_group_us[, pctl_us_cagr_20_apr_24_jul := as.integer(
  ceiling(frank(cagr_20_apr_24_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_cagr_24_jul_25_nov := as.integer(
  ceiling(frank(cagr_24_jul_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_cagr_25_jul_25_nov := as.integer(
  ceiling(frank(cagr_25_jul_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_cagr_20_apr_25_jul := as.integer(
  ceiling(frank(cagr_20_apr_25_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]
hu_block_group_us[, pctl_us_cagr_20_apr_25_nov := as.integer(
  ceiling(frank(cagr_20_apr_25_nov.x, ties.method = "min", na.last = "keep") / .N * 100)
)]

# Percentiles for agr_* metrics
hu_block_group_us[, pctl_us_agr_24_jul_25_jul := as.integer(
  ceiling(frank(agr_24_jul_25_jul.x, ties.method = "min", na.last = "keep") / .N * 100)
)]

hu_block_group_us <- hu_block_group_us[, .(
  block_group,
  idx_us_hgi_20_apr_24_jul,
  idx_us_hgi_24_jul_25_jul,
  idx_us_hgi_24_jul_25_nov,
  idx_us_hgi_25_jul_25_nov,
  idx_us_hgi_20_apr_25_jul,
  idx_us_hgi_20_apr_25_nov,
  idx_us_cagr_20_apr_24_jul,
  idx_us_cagr_24_jul_25_nov,
  idx_us_cagr_25_jul_25_nov,
  idx_us_cagr_20_apr_25_jul,
  idx_us_cagr_20_apr_25_nov,
  idx_us_agr_24_jul_25_jul,
  pctl_us_hgi_20_apr_24_jul,
  pctl_us_hgi_24_jul_25_jul,
  pctl_us_hgi_24_jul_25_nov,
  pctl_us_hgi_25_jul_25_nov,
  pctl_us_hgi_20_apr_25_jul,
  pctl_us_hgi_20_apr_25_nov,
  pctl_us_cagr_20_apr_24_jul,
  pctl_us_cagr_24_jul_25_nov,
  pctl_us_cagr_25_jul_25_nov,
  pctl_us_cagr_20_apr_25_jul,
  pctl_us_cagr_20_apr_25_nov,
  pctl_us_agr_24_jul_25_jul
)]


### WRITE TO DUCKDB

## CLEANUP FIX FOR DUCKDB BARFING DUE TO UTF ISSUES
## HAT TIP TO https://github.com/duckdb/duckdb-r/issues/12#issuecomment-2419681433
## FOR THESE STRINGI FUNCTIONS

hu_block_group_state <- hu_block_group_state |> mutate(across(where(is.character), stringi::stri_enc_tonative))
hu_block_group_county <- hu_block_group_county |> mutate(across(where(is.character), stringi::stri_enc_tonative))
hu_block_group_cbsa <- hu_block_group_cbsa |> mutate(across(where(is.character), stringi::stri_enc_tonative))
hu_block_group_us <- hu_block_group_us |> mutate(across(where(is.character), stringi::stri_enc_tonative))

# us count is one more than the others
# all NAs, remove
hu_block_group_us <- hu_block_group_us[!is.na(block_group), ]

length(unique(hu_block_group_county$block_group))
length(unique(hu_block_group_cbsa$block_group))
length(unique(hu_block_group_state$block_group))
length(unique(hu_block_group_us$block_group))

dbWriteTable(conh, "hu_block_group_county", hu_block_group_county, overwrite = TRUE)
dbWriteTable(conh, "hu_block_group_cbsa", hu_block_group_cbsa, overwrite = TRUE)
dbWriteTable(conh, "hu_block_group_state", hu_block_group_state, overwrite = TRUE)
dbWriteTable(conh, "hu_block_group_us", hu_block_group_us, overwrite = TRUE)

dbListTables(conh)

File: inprocess/bg-housing-rollups.py
## ZCTA TO STATE; 
## CORE ADAPTED FROM 'zcta-to-county.r'

import polars as pl
import duckdb
from pathlib import Path
import os
from dotenv import load_dotenv

# Set working directory
os.chdir("/home/joel")

# Load environment variables
load_dotenv()

### DUCKDB CONNECTION

conh = duckdb.connect('./data/plhousing.duckdb')
print(conh.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())

conmeta = duckdb.connect('./data/duckdb_metadata.duckdb')
print(conmeta.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())

congref = duckdb.connect('./data/georeference.duckdb', read_only=False)
print(congref.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())

congeo = duckdb.connect(
    './data/spatial_storage.duckdb',
    read_only=False
)
congeo.execute("INSTALL spatial; LOAD spatial;")

print(congeo.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())

metadata = pl.from_pandas(conmeta.execute("SELECT * FROM column_metadata").df())
tables = pl.from_pandas(conmeta.execute("SELECT * FROM table_summary").df())

# Load block groups
hu_block_group = pl.from_pandas(conh.execute("SELECT * FROM hu_block_group").df())

# Match block groups to cbsa via county
county_to_cbsa = pl.read_csv(
    "./data/county-cbsa-lookup.csv",
    dtypes={"county": pl.Utf8},
    encoding="latin1"
)

## Write to georeference duckdb
congref.execute("DROP TABLE IF EXISTS county_to_cbsa")
congref.register("county_to_cbsa", county_to_cbsa)
congref.execute("CREATE TABLE county_to_cbsa AS SELECT * FROM county_to_cbsa")
congref.unregister("county_to_cbsa")

### THIS IS A 1:1 ZCTA TO STATE CROSSWALK; WE WILL USE THIS TO ROLLUP 
### FOR BENCHMARKS OF ZCTAS COMPARED TO PARENT STATE, AND THE US

## Get hu_state, hu_zcta, etc.

hu_block_group = pl.from_pandas(conh.execute("SELECT * FROM hu_block_group").df())
hu_county = pl.from_pandas(conh.execute("SELECT * FROM hu_county").df())
hu_cbsa = pl.from_pandas(conh.execute("SELECT * FROM hu_cbsa").df())
hu_state = pl.from_pandas(conh.execute("SELECT * FROM hu_state").df())
hu_us = pl.from_pandas(conh.execute("SELECT * FROM hu_us").df())

#################################### COUNTY LEVEL BG ROLLUP #############

county_to_cbsa = pl.from_pandas(congref.execute("SELECT * FROM county_to_cbsa").df())

## Compute county fips
hu_block_group = hu_block_group.with_columns(
    pl.col("block_group").str.slice(0, 5).alias("county_fips")
)

hu_block_group_cbsa = hu_block_group.join(
    county_to_cbsa,
    left_on="county_fips",
    right_on="county",
    how="left"
)

hu_block_group_cbsa = hu_block_group_cbsa.join(
    hu_cbsa,
    on="cbsa23",
    how="left",
    suffix="_cbsa"
)

############### RELATIVE INDEX COMPUTATIONS -- CBSA LEVEL

hu_block_group_cbsa = hu_block_group_cbsa.with_columns([
    ((pl.col("hgi_20_24") - 1) / (pl.col("hgi_20_24_cbsa") - 1) * 100).alias("idx_20_24_cbsa"),
    ((pl.col("hgi_20_25") - 1) / (pl.col("hgi_20_25_cbsa") - 1) * 100).alias("idx_20_25_cbsa"),
    (pl.col("agr_25") / pl.col("agr_25_cbsa") * 100).alias("idx_25_cbsa")
])

## NULL CBSA ARE CONNECTICUT BLOCK GROUPS; CBSA DEFINITIONS BASED ON NEW PLANNING REGION
## COMPONENTS ARE NOT YET DEFINED; WE ARE USING NEW PLANNING REGION IDS IN OUR BLOCK GROUP
## IDENTIFIERS; WE WILL NEED A PLANNING REGION TO CBSA CROSSWALK TO HANDLE THESE; THESE ROWS 
## WOULD REPLACE COUNTY-BASED DEFINITIONS

hu_block_group_cbsa = hu_block_group_cbsa.filter(pl.col("block_group").is_not_null())

hu_block_group_cbsa = hu_block_group_cbsa.with_columns(
    pl.when(pl.col("cbsa23").is_null())
    .then(pl.lit("CT000"))
    .otherwise(pl.col("cbsa23"))
    .alias("cbsa23")
)

## TRIM FILE

hu_block_group_cbsa = hu_block_group_cbsa.select([
    "block_group",
    "cbsa23",
    "idx_20_24_cbsa",
    "idx_20_25_cbsa",
    "idx_25_cbsa"
])

#################################### COUNTY LEVEL BG ROLLUP #############

hu_block_group_county = hu_block_group.join(
    hu_county,
    left_on="county_fips",
    right_on="co_fips",
    how="left",
    suffix="_county"
)

################# RELATIVE INDEX COMPUTATIONS -- COUNTY LEVEL

hu_block_group_county = hu_block_group_county.with_columns([
    ((pl.col("hgi_20_24") - 1) / (pl.col("hgi_20_24_county") - 1) * 100).alias("idx_20_24_county"),
    ((pl.col("hgi_20_25") - 1) / (pl.col("hgi_20_25_county") - 1) * 100).alias("idx_20_25_county"),
    (pl.col("agr_25") / pl.col("agr_25_county") * 100).alias("idx_25_county")
])

hu_block_group_county = hu_block_group_county.filter(pl.col("block_group").is_not_null())

## TRIM FILE

hu_block_group_county = hu_block_group_county.select([
    "block_group",
    "county_fips",
    "idx_20_24_county",
    "idx_20_25_county",
    "idx_25_county"
])

################################## STATE LEVEL BG ROLLUP #############

hu_block_group = hu_block_group.with_columns(
    pl.col("block_group").str.slice(0, 2).alias("state_fips")
)

hu_block_group_state = hu_block_group.join(
    hu_state,
    left_on="state_fips",
    right_on="state_code",
    how="left",
    suffix="_state"
)

############### RELATIVE INDEX COMPUTATIONS -- STATE LEVEL

hu_block_group_state = hu_block_group_state.with_columns([
    ((pl.col("hgi_20_24") - 1) / (pl.col("hgi_20_24_state") - 1) * 100).alias("idx_20_24_state"),
    ((pl.col("hgi_20_25") - 1) / (pl.col("hgi_20_25_state") - 1) * 100).alias("idx_20_25_state"),
    (pl.col("agr_25") / pl.col("agr_25_state") * 100).alias("idx_25_state")
])

hu_block_group_state = hu_block_group_state.filter(pl.col("block_group").is_not_null())

## TRIM FILE

hu_block_group_state = hu_block_group_state.select([
    "block_group",
    "state_fips",
    "idx_20_24_state",
    "idx_20_25_state",
    "idx_25_state"
])

################################## US LEVEL BG ROLLUP #############

hu_block_group = hu_block_group.with_columns(pl.lit(1).alias("matchid"))
hu_us = hu_us.with_columns(pl.lit(1).alias("matchid"))

hu_block_group_us = hu_block_group.join(
    hu_us,
    on="matchid",
    how="left",
    suffix="_us"
)

############# RELATIVE INDEX COMPUTATIONS -- US LEVEL

hu_block_group_us = hu_block_group_us.with_columns([
    ((pl.col("hgi_20_24") - 1) / (pl.col("hgi_20_24_us") - 1) * 100).alias("idx_20_24_us"),
    ((pl.col("hgi_20_25") - 1) / (pl.col("hgi_20_25_us") - 1) * 100).alias("idx_20_25_us"),
    (pl.col("agr_25") / pl.col("agr_25_us") * 100).alias("idx_25_us")
])

### FINALLY: COMPUTE NATIONAL PERCENTILES FOR BLOCK GROUPS

# Compute percentile column for each growth index
# where the highest percentiles are the highest values
hu_block_group_us = hu_block_group_us.with_columns([
    (pl.col("hgi_20_24").rank(method="min") / pl.col("hgi_20_24").count() * 100)
        .ceil().cast(pl.Int64).alias("pctl_20_24_us"),
    (pl.col("hgi_20_25").rank(method="min") / pl.col("hgi_20_25").count() * 100)
        .ceil().cast(pl.Int64).alias("pctl_20_25_us"),
    (pl.col("agr_25").rank(method="min") / pl.col("agr_25").count() * 100)
        .ceil().cast(pl.Int64).alias("pctl_25_us")
])

hu_block_group_us = hu_block_group_us.select([
    "block_group",
    "idx_20_24_us",
    "idx_20_25_us",
    "idx_25_us",
    "pctl_20_24_us",
    "pctl_20_25_us",
    "pctl_25_us"
])

### WRITE TO DUCKDB

# Merge all indexes together
hu_block_group_indexes = hu_block_group_cbsa.join(
    hu_block_group_county, on="block_group", how="full", coalesce=True
).join(
    hu_block_group_state, on="block_group", how="full", coalesce=True
).join(
    hu_block_group_us, on="block_group", how="full", coalesce=True
)

hu_block_group_indexes = hu_block_group_indexes.filter(pl.col("block_group").is_not_null())

conh.execute("DROP TABLE IF EXISTS hu_block_group_indexes")
conh.register("hu_block_group_indexes", hu_block_group_indexes)
conh.execute("CREATE TABLE hu_block_group_indexes AS SELECT * FROM hu_block_group_indexes")
conh.unregister("hu_block_group_indexes")

#### BLOCK GROUP INDEXES COMPLETE ###############################

print(congeo.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())

block_group_names = pl.from_pandas(congeo.execute("SELECT * FROM geo_block_group").df())
block_group_hu = pl.from_pandas(conh.execute("SELECT * FROM hu_block_group").df())
block_group_indexes = pl.from_pandas(conh.execute("SELECT * FROM hu_block_group_indexes").df())

block_group_hu = block_group_names.select(["GEOID", "NAMELSAD", "STATEFP"]).join(
    block_group_hu,
    left_on="GEOID",
    right_on="block_group",
    how="left"
)

block_group_indexes = block_group_hu.join(
    block_group_indexes,
    left_on="GEOID",
    right_on="block_group",
    how="left"
)

# Filter out null HU_20
block_group_indexes = block_group_indexes.filter(pl.col("HU_20").is_not_null())

# Filter for Minnesota (STATEFP == '27')
block_group_mn = block_group_indexes.filter(pl.col("STATEFP") == "27")
print(block_group_mn)

# Close connections
conh.close()
conmeta.close()
congref.close()
congeo.close()

print("Processing complete!")

